### PR TITLE
Leaves - Alice

### PR DESCRIPTION
**Pull request by sun-alice**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model contains the state of the application and the data. 
Describe in your own words what the Controller is doing in Rails | The Controller interacts with both the view and the model, and provides the view with information from the model. 
Describe in your own words what the View is doing in Rails | The View takes the data and turns it into HTML. 
Describe an edge-case controller test you wrote | If an invalid id is provided, that the page is redirected to the home page. 
What is the purpose of using strong params? (i.e. the params method in the controller) | The purpose of strong params is to prevent unwanted parameters/data to be sent through. It sets an explicit set of attributes that may be passed through. 
How are Rails migrations related to Rails models? | Migrations change the database schemas and how the database tables are organized, which then change the way that the model gets data from the database. 
Describe one area of Rails that are still unclear on | Is the Model just supposed to represent the data in the application? Or in future projects will we do more with Models?

### CODE REVIEW

File: config/routes.rb [hunk 1]
```diff
@@ -0,0 +1,6 @@
+Rails.application.routes.draw do
+  root 'tasks#index'
+  
+  patch '/tasks/:id/complete', to: 'tasks#complete', as: 'complete_task'
+  resources :tasks
```
Comment (dHelmgren) [inline config/routes.rb:5]: Good move putting this route ahead of the rest!

File: app/views/tasks/index.html.erb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+      <p><%= link_to "Add Task", new_task_path %></p>
+      <p><%= link_to "All Tasks", tasks_path %></p>
+    </section>
+
+    <ol>
+      <h2>Your Tasks:</h2>
+      <% @tasks.each do |task| %>
+        <li>
+            <%if task.completion_date != nil %> 
+              <s><%= link_to task.name, task_path(task.id) %> </s>
```
Comment (dHelmgren) [inline app/views/tasks/index.html.erb:14]: I appreciate that you did an optional strikethru here!

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+    else
+      @task.completion_date = nil
+    end
+    
+    unless @task.save 
+      redirect_to tasks_path
+    end
+    
+    redirect_to tasks_path
+    return
```
Comment (dHelmgren) [inline app/controllers/tasks_controller.rb:90]: This is redundant since you redirect to tasks_path either way.

File: app/controllers/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+    end
+    
+    if @task.completion_date == nil
+      @task.completion_date = Time.now
+    else
+      @task.completion_date = nil
+    end
+    
+    unless @task.save 
+      redirect_to tasks_path
```
Comment (dHelmgren) [inline app/controllers/tasks_controller.rb:86]: You did well for the most part, but you forgot your `return` here

**Comment by dHelmgren** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | yes
Answered comprehension questions | yes
Successfully handles: Index, Show | yes
Index & Show tests pass | yes
Successfully handles: New, Create | yes
New & Create tests pass | yes
Successfully handles: Edit, Update | yes
Tests for Edit & Update test valid & invalid task ids | yes
Successfully handles: Destroy, Task Complete | yes
Tests for Destroy & Task Complete include tests for valid and invalid task ids | yes
Routes follow RESTful conventions | yes
Uses named routes (like `_path`) | yes
**Overall** | Good job on Task List! It functions as expected. You hit all of the learning goals: using partials, using strong params, and following Rails conventions.
